schemaregistry/schemaregistry.md

```diff
@@ -0,0 +1,257 @@
+as Apache Avro and JSON Schema.
+
+## Status of this document
+
+This document is a working draft.
+
+## 1. Introduction
+
+Schema-dependent serialization formats and related frameworks such as Apache
+Avro are popular in eventing and messaging scenarios. With most of the metadata
+information such as names and types of fields being held externally in schema
+documents, these frameworks are able to produce very compact on-wire
+representations of structured data that is carries as the payload of events and
+messages. While the wire-footprint savings are often very significant, the added
+cost is that of schema handling. All communicating parties who need to
+deserialize the event or message payload need access to the schema used to
+publish it.
+
+For formats that do not depend on external schemas for serialization, schemas
+might still be required in some scenarios to allow for consumers or inspecting
+intermediaries to validate the data structure's compliance with a set of rules.
+
+This specification defines a very simple API focused on storing, organizing, and
+accessing such documents. While there are very sophisticated metadata management
+solutions available as open source und under neutral governance already, the focus
+of this specification is not great sophistication but straightforward simplicity.
+
+This schema registry deals with only three top-level elements:
+
+- Schema Group: A named collection of schemas. Each group holds a logically
+  related set of schemas, typically belonging to a particular application or
+  subject to shared access control rules.
+- Schema: A document describing the structure, names, and types of some
+  structured data payload.
+- Version: A specific version of a schema document. Even though not prescribed
+  in this specification, an implementation might choose to impose compatibility
+  constraints on versions following the initial version of a schema.
+
+> NOTE: There are two fundamentally different options for how we can organize
+> multiple schema formats and this specification draft reflects both: 
+>
+> 1) In an "ideal" REST service, any data entity can be represented ("R" in
+>    REST) in multiple different data formats, which a client can request in
+>    prioritized order during content-type negotiation. If a service offers up
+>    the same entity in JSON, Protobuf, and Avro formats, there would therefore
+>    be three distinct schemas, each defining a representation of the data
+>    entity for the respective format. From the same "REST" mindset perspective,
+>    the ideal representation of this set of schemas, which differ with regards
+>    to the format they describe, but ought to be semantically identical, is to
+>    also differentiate them only by their content type and maintain them under
+>    the exact same name as if they were one schema.
+> 2) Since the above strategy is truly RESTful, but quite esoteric if you've not
```

Comment:
Can we have both? If a client does not specify a content type in the request, the implementation could choose a default. Or implementations could only support just one schema format.

That being said, I'm curious to hear if there's a real world example where there is value in offering multiple schema formats for the same logical schema. I can imagine in some cases that a more compact representation might be desirable (such as Avro's binary encoding over JSON), especially if bandwidth is a concern. But it does introduce some complexity, some of which will emerge in the implementations.

For example, if an implementation supports multiple formats for the same logical schema, how are they created atomically across the different representations? If the creation is not atomic, there's a chance the following scenario could technically happen with a schema registry that supports Avro and JSONSchema representations:

1. Consumer subscribes to event with Schema A
2. Schema registry registers JSONSchema version 2 of Schema A
3. Consumer receives event and requests Avro schema version 2
4. Consumer receives a 404
5. Schema registry registers Avro version 2 of Schema A

Looking at the proposed openapi spec, schema creation isn't specific about how this works, so I suppose an implementation could require sending all representations in the same request. But I wanted to call it out as complexity to consider.

Reply:
Thinking about my example some more, it's actually not valid in the context of cloud events, as the schema is determined by the producer of the event and a reference to the specific schema is enclosed in the cloud event metadata.

However, this did bring up another thought specifically with cloud events. If the schema format is indicated by the Content-Type, how does a cloud event consumer know which one to request? The dataschema URI wouldn't be enough. There's the datacontenttype, but that describes the encoding, not the schema format.

Reply:
It's okay not to do updates atomically, in my view. To use a schema, you'll need to have it registered. The consumer can't be getting an Avro encoded event if the schema has not been registered by the publisher, if the publisher play by the rules. I also don't want to be prescriptive about the compatibility enforcements; conventions are okay, too.

Reply:
Yup, I agree with that.

Regarding my second comment, in the context of cloud events, do we think that we need to provide more information in the cloud event about how the client requests the schema referenced by dataschemauri? If a registry hosts multiple schema types for the same logical schema at the same location and the client is required to indicate which type they want by sending Content-Type, then dataschemauri is no longer enough information for a client to request the correct schema that was used to structure/encode the event.

A cloud event does enclose the datacontenttype, but that describes the encoding of the event, not the type of the schema itself.

Reply:
> That being said, I'm curious to hear if there's a real world example where there is value in offering multiple schema formats for the same logical schema.

I am also very interested in a real-world example of this requirement.  It's interesting, but the various schema technologies can be so different and varied ... and I haven't seen any use cases where users need the same data format described in different schema technologies.  I can image lots of schemas where doing that would either be impractical or impossible.  But on the other hand, in a polyglot environment I can see the utility...

Reply:
Copying @gunnarmorling

After hearing all the arguments here, Im going to drop the option to keep multiple formats together under one name/version for the time being to keep things simple. If we would want to do this later, the path I chose here can indeed be added later without breaking backwards compatibility because you would just omit the now-required format declarations, which would then become optional.